include/util/cell_storage.hpp

```diff
@@ -0,0 +1,295 @@
+                               num_destination_nodes};
+            return std::make_pair(begin, end);
+        }
+
+        auto GetSourceNodes() const
+        {
+            return std::make_pair(source_boundary, source_boundary + num_source_nodes);
+        }
+
+        auto GetDestinationNodes() const
+        {
+            return std::make_pair(destination_boundary,
+                                  destination_boundary + num_destination_nodes);
+        }
+
+        CellImpl(const CellData &data,
+                 WeightPtrT const all_weight,
+                 const NodeID *const all_sources,
+                 const NodeID *const all_destinations)
+            : num_source_nodes{data.num_source_nodes},
+              num_destination_nodes{data.num_destination_nodes},
+              weights{all_weight + data.weight_offset},
+              source_boundary{all_sources + data.source_boundary_offset},
+              destination_boundary{all_destinations + data.destination_boundary_offset}
+        {
+        }
+    };
+
+    inline std::size_t LevelIDToIndex(LevelID level) const { return level-1; }
+
+  public:
+    using Cell = CellImpl<EdgeWeight *>;
+    using ConstCell = CellImpl<const EdgeWeight *>;
+
+    template <typename GraphT>
+    CellStorage(const MultiLevelPartition &partition, const GraphT &base_graph)
+    {
+        // pre-allocate storge for CellData so we can have random access to it by cell id
+        unsigned number_of_cells = 0;
+        for (LevelID level = 1u; level < partition.GetNumberOfLevels(); ++level)
+        {
+            level_to_cell_offset.push_back(number_of_cells);
+            number_of_cells += partition.GetNumberOfCells(level);
+        }
+        level_to_cell_offset.push_back(number_of_cells);
+        cells.resize(number_of_cells);
+
+        for (LevelID level = 1u; level < partition.GetNumberOfLevels(); ++level)
+        {
+            auto level_offset = level_to_cell_offset[LevelIDToIndex(level)];
+
+            std::vector<std::pair<CellID, NodeID>> level_source_boundary;
+            std::vector<std::pair<CellID, NodeID>> level_destination_boundary;
+
+            for (auto node = 0u; node < base_graph.GetNumberOfNodes(); ++node)
+            {
+                const CellID cell_id = partition.GetCell(level, node);
+                bool is_source_node = false;
+                bool is_destination_node = false;
+                bool is_boundary_node = false;
+
+                for (auto edge = base_graph.BeginEdges(node); edge < base_graph.EndEdges(node);
+                     ++edge)
+                {
+                    auto other = base_graph.GetTarget(edge);
+                    const auto &data = base_graph.GetEdgeData(edge);
+
+                    is_boundary_node |= partition.GetCell(level, other) != cell_id;
+                    is_source_node |= partition.GetCell(level, other) == cell_id && data.forward;
```

Comment:
What we are checking here is whether this node will have a row /column in the cell that could contain anything other than `INVALID_EDGE_WEIGHT`. This is only the case if there is any path starting from 0 in the cell, that could eventually lead to another border node.

There is no path in the blue cell, leading from 1 to 0. So if you would add 0 as a destination that would be a column that all weights are set to `INVALID_EDGE_WEIGHT`. However we there is an edge going from 0 to 1 so there could be a path to another border vertex (also `1` in this case).